[#5392] disable "sign transaction" button when offline

dmitryn opened the pull request:
fixes #5392

### Summary:

Disable "sign transaction" button when offline.

status: ready

dmitryn committed:
[#5392] disable "sign transaction" button when offline

Signed-off-by: Dmitry Novotochinov <dmitry.novot@gmail.com>

flexsurfer commented on src/status_im/ui/screens/wallet/send/views.cljs:
```diff
@@ -106,18 +107,22 @@
 ;; "Sign Transaction >" button
 (defn- sign-transaction-button [amount-error to amount sufficient-funds? sufficient-gas? modal?]
-  (let [sign-enabled? (and (nil? amount-error)
+  (let [online? (= :online @(re-frame/subscribe [:network-status]))
```
this looks weird, can we pass online? from outside?

dmitryn replied:
@flexsurfer changed, please take a look

asemiankevich commented:
@dmitryn `transaction underpriced` is received in the following case:

- open status
- faucet some funds
- open 1x1 chat
- go offline
- tap send ETH 0 and go to transaction screen. Note that `sign transaction` button is disabled
- go online (turn off airplane mode)
- check that sign button is available
- tap sign transaction

as a result, user sees `transaction underpriced` pop-up while expecting the transaction is sent. 

Seems gas limit / price are not fetched correctly (i do see values in UI but seems they are not taken into account)

dmitryn commented:
@asemiankevich i've hidden submit button for /send /receive command when offline, please take a look.

lukaszfryc commented:
@dmitryn I could not reproduce Nastya's issue. But, I noticed 2 other things that maybe related to this PR.

I had a poor network connection and couple of times I experienced:
1. `t.minus` error while sending ETH in 1-1 chat

<img src="https://user-images.githubusercontent.com/7532782/45613803-bb34d600-ba67-11e8-959f-73028f83457c.png" width="200">

I could not found TF session for it.

2. 0 gas price even when my network connection returned back:

<img src="https://user-images.githubusercontent.com/7532782/45613841-d56eb400-ba67-11e8-8d63-4c92a7f0d247.png" width="200">

[TF](https://app.testfairy.com/projects/4803590-status/builds/8594906/sessions/4401667113/?accessToken=OlCsh8hjn3k/L08u9UK1MCfHtMA) for 0 gas.


Do you think those issues are related?

lukaszfryc commented:
@dmitryn I actually think those issues are not related to your change. I see them in other PRs as well.

lukaszfryc commented:
@dmitryn but `TypeError: null is not an object (evaluating 't.minus')` can still be replicated by following these steps:

- Create account, add a contact, switch to ropsten and request ETH
- Kill the app 
- Disable Wi-Fi and Mobile data
- Open Status
- Log into the existing account
- Go to Wallet and send transaction
- Select the contact
- Try to type some ETH amount 

-> The error appears. I think it maybe related to the fact that wallet and send transactions do not have amount, gas limit, etc. fetched because the device was offline. Anyway, this error should not appear.

dmitryn commented:
@lukaszfryc i've disabled amount input when offline to fix issue you found. Please take a look.

lukaszfryc commented:
@dmitryn 

1) When I followed steps from my recent comment https://github.com/status-im/status-react/pull/5817#issuecomment-422021405, opened wallet and went through wallet onboarding I got this error:

<img src="https://user-images.githubusercontent.com/7532782/45674739-01546d00-bb2e-11e8-91ed-eb4d3bf28208.png" width="200">


2) Another thing, the amount is disabled in send transaction screen as you mentioned. But, I strongly believe that we should inform user why we disabled it. How about having a gray bar as we have in Wallet saying "Can't send transaction in offline mode":

<img src="https://user-images.githubusercontent.com/7532782/45674513-94d96e00-bb2d-11e8-80c5-36e3374472be.png" width="200">

<img src="https://user-images.githubusercontent.com/7532782/45674679-dff38100-bb2d-11e8-8e9a-48717eb3015a.png" width="200">

dmitryn commented:
@lukaszfryc please take a look.
1. fixed
2. good idea, added snackbar with that message when offline.

lukaszfryc commented:
> good idea, added snackbar with that message when offline.

1) The snackbar looks is really nice! Can you add it to "Sign Message" screen as well? Copy can say "Can't sign message in offline mode"

2) I got `null is not an object (evaluating 't.minus')` once again after I installed the app, switched to ropsten requested eth in test dapp. Here is [TF, see from 4:30](https://app.testfairy.com/projects/4803622-status/builds/8596958/sessions/4401762042/?accessToken=bUHwcVnuR3oLKLu1CKjRdxRyChQ). I could not replicate it afterwards :/ 

```
09-18 15:48:49.567 15566 15702 E ReactNativeJS: TypeError: null is not an object (evaluating 't.minus')
09-18 15:48:49.567 15566 15702 E ReactNativeJS:
09-18 15:48:49.567 15566 15702 E ReactNativeJS: This error is located at:
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in send-transaction-modal
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent115
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent14
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main-screen-modal-view
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent110
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTModalHostView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in t
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main-modal
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent3
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in root
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent1
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in t
```

dmitryn commented:
@lukaszfryc snackbar added and error fixed (hopefully). Please take a look.

lukaszfryc commented:
@dmitryn 

> I got null is not an object (evaluating 't.minus') once again after I installed the app, switched to ropsten requested eth in test dapp. Here is TF, see from 4:30. I could not replicate it afterwards :/

I could not reproduce it so I think it's fixed now 👍 

----

But, there are still to do:

> The snackbar looks is really nice! Can you add it to "Sign Message" screen as well? Copy can say "Can't sign message in offline mode"

1) Looks like there is no copy for sign message bar. 

<img src="https://user-images.githubusercontent.com/7532782/45749132-0cc99600-bc0b-11e8-9c87-06805ea44df4.png" width="200">

2) On iPhone, the "Amount" field is not clickable after app goes back online. To reproduce go to sign transaction, go offline and then back online. The amount field should be clickable now but it is not.

<img src="https://user-images.githubusercontent.com/7532782/45749403-e1937680-bc0b-11e8-9d6b-20a70acb1f41.png" width="200">

dmitryn commented:
@lukaszfryc 

1. added translation
2. There seems a bug when with returning from background on IOS – app state doesn't change back to "online" (i've tested it in develop and have same issue, i even see "Can't send when offline" snackbar when i return back online. It'll be helpful if you could reproduce it on real ios device in develop branch). 
So i switched to not disabling amount field when offline, this doesn't give my any errors and field is still editable when online.

Cannot make build at the moment (jenkins issue), will ping you when new build will be available.

dmitryn commented:
@lukaszfryc new build is ready, please check.

lukaszfryc commented:
@dmitryn I'll come back to it later today once [release blockers](https://github.com/status-im/status-react/labels/release) are fixed.

lukaszfryc commented:
@dmitryn [latest bugs](https://github.com/status-im/status-react/pull/5817#issuecomment-422788765) are fixed but there is a new one. Hopefully the last one as I could not find anything else :) 

On Android (I'm testing on 8.0), when you **device press back button** in transaction fee screen, there is an error:

<img src="https://user-images.githubusercontent.com/7532782/45805012-f4668380-bcbd-11e8-8365-81e3927e3524.png" width="200">

<img src="https://user-images.githubusercontent.com/7532782/45805042-034d3600-bcbe-11e8-9c24-c306c888b7ed.png" width="200">

This only happen when back button is pressed. Works fine for "X" or "Done" buttons. 

I checked latest nightly (`ce741d7`) and all works fine there so I assume it's related to our PR.

dmitryn commented:
@lukaszfryc i've rebased branch to develop and this error is gone. Waiting for Jenkins build.

dmitryn commented:
@lukaszfryc build is ready, please check.

lukaszfryc commented:
I’m on it.

lukaszfryc commented:
@dmitryn I did not notice any regressions or new issues. Feel free to merge 💯

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

lukaszfryc commented:
@dmitryn please cherry pick it to `release/0.9.28` as well.

dmitryn commented:
@lukaszfryc sure, cherry-picked to release/0.9.28. Thanks for testing!
